[PyTorch] Devirtualize TensorImpl::has_storage

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51050 [PyTorch] Devirtualize TensorImpl::storage()
* #50920 [PyTorch] Re-order TensorImpl fields to save a word
* **#51049 [PyTorch] Devirtualize TensorImpl::has_storage**
* #51048 [PyTorch] Devirtualize TensorImpl::storage_offset

This diff makes it OK to query has_storage() on all TensorImpls. I added debug assertions that storage_ is indeed never set on them, which is required for this to be correct.

Differential Revision: [D26008498](https://our.internmc.facebook.com/intern/diff/D26008498/)

swolchok committed:
[PyTorch] Devirtualize TensorImpl::has_storage

This diff makes it OK to query has_storage() on all TensorImpls. I added debug assertions that storage_ is indeed never set on them, which is required for this to be correct.

Differential Revision: [D26008498](https://our.internmc.facebook.com/intern/diff/D26008498/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Devirtualize TensorImpl::has_storage"


This diff makes it OK to query has_storage() on all TensorImpls. I added debug assertions that storage_ is indeed never set on them, which is required for this to be correct.

Differential Revision: [D26008498](https://our.internmc.facebook.com/intern/diff/D26008498/)

[ghstack-poisoned]

ezyang reviewed (approved):
I, for one, vote for just straight deletion, but this is OK too.

bhosmer commented:
> I, for one, vote for just straight deletion, but this is OK too.

@ezyang is straight deletion an option? That would certainly leave things in a much simpler state, but do existing callsites all vanish cleanly?

swolchok commented:
> > I, for one, vote for just straight deletion, but this is OK too.
> 
> @ezyang is straight deletion an option? That would certainly leave things in a much simpler state, but do existing callsites all vanish cleanly?

I think callers can just call `storage()` and see if they get a truthy Storage object.

swolchok committed:
Update on "[PyTorch] Devirtualize TensorImpl::has_storage"


This diff makes it OK to query has_storage() on all TensorImpls. I added debug assertions that storage_ is indeed never set on them, which is required for this to be correct.

Differential Revision: [D26008498](https://our.internmc.facebook.com/intern/diff/D26008498/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Devirtualize TensorImpl::has_storage"


This diff makes it OK to query has_storage() on all TensorImpls. I added debug assertions that storage_ is indeed never set on them, which is required for this to be correct.

Differential Revision: [D26008498](https://our.internmc.facebook.com/intern/diff/D26008498/)

[ghstack-poisoned]
